File: java/allergies/src/main/java/Allergies.java
import java.util.List;
import java.util.ArrayList;

class Allergies {
  int score;

  Allergies(int score) {
    this.score = score;
  }

  public boolean isAllergicTo(Allergen a) {
    if ((this.score & a.getScore()) != 0) {
      return true;
    }
    return false;
  }

  public List<Allergen> getList() {
    List<Allergen> list = new ArrayList<Allergen>();
    for (Allergen a : Allergen.values()) {
      if (isAllergicTo(a)) {
        list.add(a);
      }
    }
    return list;
  }
}

File: c/allergies/src/allergies.c
#include "allergies.h"
#include <stdlib.h>

bool is_allergic_to(allergen_t allergen, int score)
{
  return allergen & score;
}

void get_allergens(int score, allergen_list_t* allergen_list)
{
  allergen_list->allergens = malloc(ALLERGEN_COUNT * sizeof(allergen_t));
  allergen_list->count = 0;

  for (allergen_t allergen = ALLERGEN_EGGS; allergen <= ALLERGEN_CATS; allergen *= 2)
  {
    if (score & allergen)
    {
      allergen_list->allergens[allergen_list->count] = allergen;
      allergen_list->count++;
    }
  }
}
